crates/matrix-sdk/src/widget/mod.rs

```diff
@@ -50,6 +42,89 @@ pub struct Comm {
+    /// This contains the url from the widget state event.
+    /// In this url placeholders can be used to pass information from the client
+    /// to the widget. Possible values are: `$widgetId`, `$parentUrl`,
+    /// `$userId`, `$lang`, `$fontScale`, `$analyticsID`.
+    ///
+    /// # Examples
+    ///
+    /// e.g `http://widget.domain?username=$userId`
+    /// will become: `http://widget.domain?username=@user_matrix_id:server.domain`.
+    raw_url: Url,
+}
+impl WidgetSettings {
+    /// Create the actual Url that can be used to setup the WebView or IFrame
+    /// that contains the widget.
+    ///
+    /// # Arguments
+    ///
+    /// * `room` - A matrix room which is used to query the logged in username
+    /// * `parent_url` - The parent url is used as the target for the
+    ///   postMessages send by the widget
+    /// Should be the url of the app hosting the widget.
+    /// In case the app hosting the widget is not a webapp the platform specific
+    /// value needs to be used or `"*"` a wildcard.
+    /// Be aware that this means the widget will receive its own postmessage
+    /// messages. The (js) matrix-widget-api ignores those however so this
+    /// works but it might break custom implementations. So always keep this
+    /// in mind.
+    /// * `font_scale` - The font scale used in the widget.
+    /// This should be in sync with the current client app configuration
+    /// * `lang` - the language e.g. en-us
+    /// * `analytics_id` - This can be used in case the widget wants to connect
+    ///   to the
+    /// same analytics provider as the client app only set this value on widgets
+    /// which are known.
+    pub fn get_url(
+        &self,
+        room: JoinedRoom,
+        parent_url: &str,
+        font_scale: f64,
+        lang: LanguageTag,
+        analytics_id: &str,
+    ) -> String {
+        self.raw_url
+            .as_str()
+            .replace("$widgetId", &self.id)
+            .replace("$parentUrl", &encode(parent_url))
+            .replace("$userId", room.own_user_id().as_str())
+            .replace("$lang", lang.as_str())
+            .replace("$fontScale", &font_scale.to_string())
+            .replace("$analyticsID", analytics_id)
+    }
+
+    /// WidgetSettings are usually created from a state event.
+    /// (currently unimplemented)
+    /// But in some cases the client wants to create custom `WidgetSettings`
+    /// for specific rooms based on other conditions.
+    /// This function returns a WidgetSettings object which can be used
+    /// to setup a widget using `run_client_widget_api`
+    /// and to generate the correct url for the widget.
+    pub fn new_virtual_widget(
+        id: String,
+        init_on_load: bool,
+        base_path: &str,
+    ) -> Result<Self, url::ParseError> {
+        let raw_url = format!("{base_path}?widgetId=$widgetId&parentUrl=$parentUrl#?&userId=$userId&lang=$lang&fontScale=$fontScale&analyticsID=$analyticsID", base_path= base_path);
```

Comment:
```suggestion
        let raw_url = format!("{base_path}?widgetId=$widgetId&parentUrl=$parentUrl#?&userId=$userId&lang=$lang&fontScale=$fontScale&analyticsID=$analyticsID");
```

If the capturing name has the same name as a variable, it automatically capture the variable.